spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/ThreadAffinityClientConnectionFactory.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.ip.tcp.connection;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.core.convert.ConversionService;
+import org.springframework.core.serializer.Deserializer;
+import org.springframework.core.serializer.Serializer;
+import org.springframework.expression.Expression;
+import org.springframework.integration.support.MessageBuilderFactory;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.core.DestinationResolver;
+import org.springframework.util.Assert;
+
+/**
+ * A client connection factory that binds a connection to a thread. Close operations
+ * are ignored; to physically close a connection and release the thread local, invoke
+ * {@link #releaseConnection()).
+ *
+ * @author Gary Russell
+ * @since 5.0
+ *
+ */
+public class ThreadAffinityClientConnectionFactory extends AbstractClientConnectionFactory {
+
+	private final AbstractClientConnectionFactory connectionFactory;
+
+	/*
+	 * Not static because we might have several factories with different delegates.
+	 */
+	private final ThreadLocal<TcpConnectionSupport> connections = new ThreadLocal<>();
+
+	public ThreadAffinityClientConnectionFactory(AbstractClientConnectionFactory connectionFactory) {
+		super("", 0);
+		Assert.isTrue(connectionFactory.isSingleUse(),
+				"ConnectionFactory must be single-use to assign a connection per thread");
+		this.connectionFactory = connectionFactory;
+	}
+
+	public void releaseConnection() {
+		TcpConnectionSupport connection = this.connections.get();
+		if (connection != null) {
+			this.connections.remove();
+			connection.enableClose();
+			connection.close();
+		}
+	}
+
+	@Override
+	public TcpConnectionSupport getConnection() throws Exception {
+		TcpConnectionSupport connection = this.connections.get();
+		if (connection == null || !connection.isOpen()) {
+			connection = this.connectionFactory.getConnection();
+			connection.disableClose();
+			this.connections.set(connection);
+		}
+		return connection;
```

Comment:
Can't we instead of that peculiar `disable/enableClose` return here a wrapper with the empty `close()` impl and invoke `delegate.close()` in that `releaseConnection()` above?

Something similar we have in the Spring Kafka: https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java#L108